File: csharp/DataStructures/src/dsa/search/LinearSearch.cs
namespace dsa.search;

public class LinearSearch
{
    // static void Main()
    // {
    //     int[] arr = { 5, 1, 9, 2, 10, 15, 20 };
    //     LinearSearch ls = new();
    //     Console.WriteLine(ls.search(arr, arr.Length, 9));
    // }

    public int search(int[] arr, int n, int x)
    {
        // Verifica se o array é nulo ou vazio. Se for, lança uma exceção.
        if (arr == null || arr.Length == 0)
            throw new Exception("Invalid input");

        // Itera por cada item do array. A variável 'i' é o índice atual.
        for (int i = 0; i < n; i++)
        {
            // Compara o elemento atual do array (arr[i]) com o valor alvo (x).
            // Se eles forem iguais, retorna o índice atual, pois o valor alvo foi encontrado.
            if (arr[i] == x)
                return i;
        }

        // Se o valor alvo não foi encontrado após percorrer todo o array, retorna -1.
        return -1;
    }
}

File: go/dsa/search/LinearSearch.go
package search

import "errors"

// Search é uma função que procura por um elemento 'x' em um array 'arr'.
// Retorna o índice do elemento se encontrado, ou um erro se o array estiver vazio ou o elemento não for encontrado.
func Search(arr []int, x int) (int, error) {

	// Verifica se o array está vazio. Se estiver, retorna um erro.
	if len(arr) == 0 {
		return -1, errors.New("array is null")
	}

	// Itera sobre o array para procurar o elemento 'x'.
	for i := 0; i < len(arr); i++ {
		if arr[i] == x {
			return i, nil // Retorna o índice do elemento encontrado.
		}
	}

	// Se o elemento não for encontrado, retorna -1 e um erro informando que o elemento não foi encontrado.
	return -1, errors.New("element not found")
}
